support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/authenticator/OAuth20RefreshTokenAuthenticator.java

```diff
@@ -0,0 +1,79 @@
+import org.pac4j.core.exception.CredentialsException;
+
+import java.io.Serializable;
+
+/**
+ * This is {@link OAuth20RefreshTokenAuthenticator}.
+ *
+ * @author Julien Huon
+ * @since 6.2.0
+ */
+@Slf4j
+public class OAuth20RefreshTokenAuthenticator extends OAuth20ClientIdClientSecretAuthenticator {
+
+    public OAuth20RefreshTokenAuthenticator(final ServicesManager servicesManager,
+                                            final ServiceFactory webApplicationServiceFactory,
+                                            final AuditableExecution registeredServiceAccessStrategyEnforcer,
+                                            final TicketRegistry ticketRegistry,
+                                            final CipherExecutor<Serializable, String> registeredServiceCipherExecutor,
+                                            final PrincipalResolver principalResolver) {
+        super(servicesManager, webApplicationServiceFactory, registeredServiceAccessStrategyEnforcer,
+            registeredServiceCipherExecutor, ticketRegistry, principalResolver);
+    }
+
+    /**
+     * RefreshTokenAuthenticator can only be used for a refresh token request of a "public" client.
+     * An OAuth "public" client is one that does not define a secret like a mobile application.
+     *
+     * @param context the context
+     * @return true if authenticator can validate credentials.
+     */
+    @Override
+    protected boolean canAuthenticate(final WebContext context) {
+        val grantType = context.getRequestParameter(OAuth20Constants.GRANT_TYPE);
+        val clientId = context.getRequestParameter(OAuth20Constants.CLIENT_ID);
+
+        if (clientId.isPresent() && grantType.isPresent()
+            && OAuth20Utils.isGrantType(grantType.get(), OAuth20GrantTypes.REFRESH_TOKEN)
+            && context.getRequestParameter(OAuth20Constants.REFRESH_TOKEN).isPresent()) {
+            val registeredService = OAuth20Utils.getRegisteredOAuthServiceByClientId(getServicesManager(), clientId.get());
+
+            if (registeredService != null && !OAuth20Utils.doesServiceNeedAuthentication(registeredService)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    protected void validateCredentials(final UsernamePasswordCredentials credentials,
+                                       final OAuthRegisteredService registeredService, final WebContext context) {
+        val token = credentials.getPassword();
+        LOGGER.trace("Received refresh token [{}] for authentication", token);
+
+        val refreshToken = getTicketRegistry().getTicket(token, OAuth20RefreshToken.class);
+        if (refreshToken == null || refreshToken.isExpired() || !StringUtils.equals(refreshToken.getClientId(), credentials.getUsername())) {
+            LOGGER.error("Provided refresh token [{}] is either not found in the ticket registry or has expired or not related with the client provided");
```

Comment:
```suggestion
            LOGGER.error("Provided refresh token [{}] is either not found in the ticket registry, has expired or is not related to the client provided");
```

Reply:
Could we also include the client-id in log statement?

Reply:
Sure, will add it.